jaxtronomy/LensModel/Solver/lens_equation_solver.py

```diff
@@ -0,0 +1,773 @@
+import jax.numpy as jnp
+from jax import jit
+import numpy as np
+import jaxtronomy.Util.util as util
+import lenstronomy.Util.image_util as image_util
+from jax.scipy.optimize import minimize
+from lenstronomy.LensModel.Solver.epl_shear_solver import solve_lenseq_pemd
+
+__all__ = ["LensEquationSolver"]
+
+SUPPORTED_LENS_MODELS_ANALYTICAL = (
+    ["SIE", "SHEAR"],
+    ["SIE"],
+    ["EPL_NUMBA", "SHEAR"],
+    ["EPL_NUMBA"],
+    ["EPL", "SHEAR"],
+    ["EPL"],
+)
+
+
+class LensEquationSolver(object):
+    """Class to solve for image positions given lens model and source position."""
+
+    def __init__(self, lensModel):
+        """This class must contain the following definitions (with same syntax as the
+        standard LensModel() class: def ray_shooting() def hessian() def magnification()
+
+        :param lensModel: instance of a class according to
+            lenstronomy.LensModel.lens_model
+        """
+        self.lensModel = lensModel
+
+    # --------------------------------------------------------------------------------
+    # The following two methods are required to allow the JAX compiler to recognize
+    # the LensEquationSolver class. Methods involving the self variable can be
+    # jit-decorated, and will recompile whenever the self variables listed under
+    # aux_data are changed
+    def _tree_flatten(self):
+        children = ()
+        aux_data = {"lensModel": self.lensModel}
+        return (children, aux_data)
+
+    @classmethod
+    def _tree_unflatten(cls, aux_data, children):
+        return cls(*children, **aux_data)
+
+    # --------------------------------------------------------------------------------
+
+    def image_position_from_source(
+        self, sourcePos_x, sourcePos_y, kwargs_lens, solver="lenstronomy", **kwargs
+    ):
+        """Solves the lens equation, i.e. finds the image positions in the lens plane
+        that are mapped to a given source position.
+
+        :param sourcePos_x: source position in units of angle
+        :param sourcePos_y: source position in units of angle
+        :param kwargs_lens: lens model parameters as keyword arguments
+        :param solver: which solver to use, can be 'lenstronomy' (default), 'analytical'
+            or 'stochastic'.
+        :param kwargs: Any additional kwargs are passed to the chosen solver, see the
+            documentation of image_position_lenstronomy, image_position_analytical and
+            image_position_stochastic
+        :returns: (exact) angular position of (multiple) images ra_pos, dec_pos in units
+            of angle
+        """
+        if solver == "lenstronomy":
+            return self.image_position_lenstronomy(
+                sourcePos_x, sourcePos_y, kwargs_lens, **kwargs
+            )
+        if solver == "analytical":
+            return self.image_position_analytical(
+                sourcePos_x, sourcePos_y, kwargs_lens, **kwargs
+            )
+
+    def image_position_lenstronomy(
```

Comment:
just curious, you don't need a jit decorator here?

Reply:
Same as my comment above